### DISCUSSION TITLE
examples/quantized-qwen2-instruct:  text generation

### DISCUSSION THREAD

**Discussion by kotamine** [huggingface/candle#2856] [General]
@greenrazer  
in https://github.com/huggingface/candle/blob/main/candle-examples/examples/quantized-qwen2-instruct/main.rs,  
lines 284-286 has the following code 

`
for index in 0..to_sample {
        let input = Tensor::new(&[next_token], &device)?.unsqueeze(0)?;
        let logits = model.forward(&input, tokens.len() + index)?;
`
and I wonder if this is input specification is as intended.  It may not be feeding the tokens generated recursively.  
(and I see the similar code in quantized-phi page.) 

For example, examples/phi  https://github.com/huggingface/candle/blob/main/candle-examples/examples/phi/main.rs  has the following
`
let ctxt = &tokens[tokens.len().saturating_sub(context_size)..];
            let input = Tensor::new(ctxt, &self.device)?.unsqueeze(0)?;
`
which is appending each new token generation to `tokens` and feeding that into the next iteration in the loop.

**Comment by greenrazer** [comment]
Hi @kotamine,

It appears that while both models are using the KV-cache internally in their respective attention layers, their api's are slightly different. 

Quantized Qwen2 KV-Cache:
https://github.com/huggingface/candle/blob/b4daa03e598b516ee4dca5864b70f7254642b7bd/candle-transformers/src/models/quantized_qwen2.rs#L115

PHI KV-Cache:
https://github.com/huggingface/candle/blob/b4daa03e598b516ee4dca5864b70f7254642b7bd/candle-transformers/src/models/phi.rs#L229

The Qwen2 model seems to take a more imperitive approach to it's API, explicitly taking a position index to the forward method. Which is used for rotary position embeddings and to manage the KV cache:

https://github.com/huggingface/candle/blob/b4daa03e598b516ee4dca5864b70f7254642b7bd/candle-transformers/src/models/quantized_qwen2.rs#L112

The Phi model seems to take a more functional approach to it's API, implicitly handling position in its forward method. It infers the sequence length from the input dimensions and the size of the kv-cache:

https://github.com/huggingface/candle/blob/b4daa03e598b516ee4dca5864b70f7254642b7bd/candle-transformers/src/models/phi.rs#L346
https://github.com/huggingface/candle/blob/b4daa03e598b516ee4dca5864b70f7254642b7bd/candle-transformers/src/models/phi.rs#L217

However if you notice a bug where the output of the `quantized-qwen2` models are outputting gibberish, or tokens that don't make sense with it's context window, please open an issue!

**Comment by kotamine** [comment]
Thank you for the clarification.
Cache use makes sense now.
It’s working great as is!

**Reply by kotamine** [reply]
I'm hoping to use the model in an app and generate texts for multiple, sequential prompts as they come (instead of one-off use and loading and off-loading from GPU).   Is clearing the kv_cache between text generations the right approach?  like 
`for layer in model.layers.iter_mut() {
            layer.kv_cache = None; 
        }  
`  

also, I noticed that for the smallest model, `--features cuda`   seems to lead to a negative or too large weight error.  For example, 
`cargo run --features cuda --example quantized-qwen2-instruct --release -- --which "0.5b"`   If this type of error won't happen for larger models, I can ignore it.   

I meant to thank you for these examples. I was also able to add `Qwen2.5-Coder-14B-Instruct` models with your code as-is (since the model structures are the same)!

**Reply by greenrazer** [reply]
Hi kotamine,

For the first part, I believe that is what the `phi` models are doing:

https://github.com/huggingface/candle/blob/b4daa03e598b516ee4dca5864b70f7254642b7bd/candle-transformers/src/models/phi.rs#L363

For the second part, I don't have a cuda machine so I cannot test, but I'd suggest opening an issue to get the right eyes on it/track fixes :).
